### PR TITLE
Respect FK constraints for extension config tables

### LINKED ISSUE #701: pgcopydb doesn't respect foreign key dependencies for extension configuration tables
Extension configuration tables can have foreign key dependencies between them. If the configuration tables are restored in the incorrect order, it can cause foreign key constraint violations.

pgcopydb (implicitly?) [assumes](https://github.com/dimitri/pgcopydb/blob/ff6b47389fd68447ad9e2039a418b97fce9e3186/src/bin/pgcopydb/schema.c#L426-L450) that the ordering of the oids in `pg_extension.extconfig` is sufficient, but it is not (see pg_dump's [implementation](https://github.com/postgres/postgres/blob/539e328b1c3a4047c7ff525524edbcca2bac6e10/src/bin/pg_dump/pg_dump.c#L17858-L17880)).

## Steps to reproduce

Note: this requires the extension modifying its own catalog to get the config tables in a weird order, so these steps to reproduce must be followed exactly.

```
# setup source db
docker run -d -e POSTGRES_HOST_AUTH_METHOD=trust -p 5432:5432 timescale/timescaledb-ha:pg15

# setup target db
docker run -d -e POSTGRES_HOST_AUTH_METHOD=trust -p 5433:5432 timescale/timescaledb-ha:pg15

# load specific version (these specific versions modify the catalog in a way that illustrates this issue)
psql postgres://postgres@localhost:5432 -c "DROP EXTENSION timescaledb;"
psql postgres://postgres@localhost:5432 -c "CREATE EXTENSION timescaledb VERSION '2.11.2';"
psql postgres://postgres@localhost:5432 -c "ALTER EXTENSION timescaledb UPDATE TO '2.12.0';"

# correctly set up target
psql postgres://postgres@localhost:5432 -c "DROP EXTENSION timescaledb;"
psql postgres://postgres@localhost:5432 -c "CREATE EXTENSION timescaledb VERSION '2.12.0';"

# ensure that there's data to cause a FK violation
psql postgres://postgres@localhost:5432 -c "CREATE TABLE metrics(time timestamptz, value bigint); SELECT create_hypertable('metrics', 'time');"

# migrate
docker run --network=host -ti dimitri/pgcopydb:v0.15 pgcopydb copy extensions --source postgres://postgres@localhost:5432 --target postgres://postgres@localhost:5433 --verbose
```

Output
```
12:40:35.014 1 INFO   Running pgcopydb version 0.15 from "/usr/local/bin/pgcopydb"
12:40:35.015 1 INFO   [SOURCE] Copying database from "postgres://postgres@localhost:5432/?keepalives=1&keepalives_idle=10&keepalives_interval=10&keepalives_count=60"
12:40:35.015 1 INFO   [TARGET] Copying database into "postgres://postgres@localhost:5433/?keepalives=1&keepalives_idle=10&keepalives_interval=10&keepalives_count=60"
12:40:35.037 1 INFO   Using work dir "/tmp/pgcopydb"
12:40:35.042 1 INFO   Exported snapshot "00000004-00000011-1" from the source database
12:40:35.042 1 NOTICE Wrote snapshot "00000004-00000011-1" to file "/tmp/pgcopydb/snapshot"
12:40:35.118 1 NOTICE Registering catalog setup for source "postgres://postgres@localhost:5432/", target "postgres://postgres@localhost:5433/", snapshot "00000004-00000011-1"
12:40:35.170 1 INFO   Fetched information for 3 extensions
12:40:35.238 1 INFO   Found 0 indexes (supporting 0 constraints) in the target database
12:40:35.240 1 INFO   Creating extension "plpgsql"
12:40:35.244 1 WARN   NOTICE:  extension "plpgsql" already exists, skipping
12:40:35.244 1 INFO   Creating extension "timescaledb"
12:40:35.252 1 WARN   NOTICE:  extension "timescaledb" already exists, skipping
12:40:35.253 1 INFO   COPY extension "timescaledb" configuration table _timescaledb_catalog.hypertable_data_node
12:40:35.258 1 INFO   COPY extension "timescaledb" configuration table _timescaledb_catalog.tablespace
12:40:35.266 1 INFO   COPY extension "timescaledb" configuration sequence _timescaledb_catalog.dimension_id_seq
12:40:35.278 1 INFO   COPY extension "timescaledb" configuration table _timescaledb_catalog.dimension
12:40:35.291 1 ERROR  [TARGET 97] [23503] ERROR:  insert or update on table "dimension" violates foreign key constraint "dimension_hypertable_id_fkey"
12:40:35.291 1 ERROR  [TARGET 97] DETAIL:  Key (hypertable_id)=(1) is not present in table "hypertable".
12:40:35.291 1 ERROR  [TARGET 97] Context: Failed to copy data to target
12:40:35.291 1 ERROR  Failed to iterate over list of extensions, see above for details
12:40:35.291 1 ERROR  Failed to copy extensions, see above for details```

### PR DESCRIPTION
**Pull request by JamesGuthrie**
Extension configuration tables can have foreign key dependencies between them. If the configuration tables are restored in the incorrect order, it can cause foreign key constraint violations.

This change ensures that the configuration tables are ordered according to the dependency order in `pg_constraint`.

I don't expect that this change will be merged as-is, but I've tested it for our use-case and it seems to work.

Fixes #701

### CODE REVIEW

File: src/bin/pgcopydb/schema.c [hunk 1]
```diff
@@ -423,11 +423,20 @@ schema_list_extensions(PGSQL *pgsql, DatabaseCatalog *catalog)
...
 
 	char *sql =
+		"with recursive fk_constraints as ("
+		"     select * from pg_constraint where contype = 'f' and conrelid <> confrelid"
+		" ), ordered_fk_constraints as ("
+		"     select distinct c.confrelid as relid, 0 depth from fk_constraints c where not exists (select * from fk_constraints fc where fc.conrelid = c.confrelid)"
+		"     UNION"
+		"     select distinct c.conrelid as relid, r.depth + 1 FROM ordered_fk_constraints r join fk_constraints c ON c.confrelid = r.relid"
+		" ), min_constraint AS (select relid, min(depth) as depth FROM ordered_fk_constraints group by relid),"
+		" preprocessed as ("
```
Comment (dimitri) [inline src/bin/pgcopydb/schema.c:433]: Please reformat the SQL query for our line length limit (target < 80 for readability, allow oversize lines when you feel like it's better, up to 90 chars).

**Review by dimitri — CHANGES_REQUESTED** [review]
This PR is simple enough (only touches one small relevant part of the code) to be a good candidate for inclusion. Thanks for your contribution! Please consider the following changes for a merge.

File: src/bin/pgcopydb/schema.c [hunk 2]
```diff
@@ -4073,7 +4092,7 @@ getExtensionList(void *ctx, PGresult *result)
 
 	if (PQnfields(result) != 11)
 	{
-		log_error("Query returned %d columns, expected 10", PQnfields(result));
+		log_error("Query returned %d columns, expected 11", PQnfields(result));
```
[the remark this answers is not in the document]
Comment (JamesGuthrie) [reply src/bin/pgcopydb/schema.c:4095]: This fixes a minor bug which I ran into while developing this patch. The text of the error does not correspond with the actual check on the line above (which checks for 11, not 10). I did not change the number of fields returned by the query.

[the remark this answers is not in the document]
Comment (dimitri) [reply src/bin/pgcopydb/schema.c:4095]: Oh yeah, I misread that. Sure, makes sense. Thanks!

File: src/bin/pgcopydb/schema.c (hunk 1 above)
Comment (JamesGuthrie) [reply src/bin/pgcopydb/schema.c]: Done

File: src/bin/pgcopydb/schema.c [hunk 3]
```diff
@@ -423,11 +423,34 @@ schema_list_extensions(PGSQL *pgsql, DatabaseCatalog *catalog)
 	SourceExtensionArrayContext parseContext = { { 0 }, catalog, false };
 
 	char *sql =
+		"with recursive fk_constraints as ( "
+		"     select * from pg_constraint where contype = 'f' and conrelid <> confrelid "
```
Comment (dimitri) [inline src/bin/pgcopydb/schema.c:427]: Forgot to mention: `SELECT *` is considered harmful in code. Interactive use is another story. Please expand the column names you need. Hint: we don't know what version of Postgres the code is going to connect to at run-time.

File: src/bin/pgcopydb/schema.c [hunk 4]
```diff
@@ -423,11 +423,34 @@ schema_list_extensions(PGSQL *pgsql, DatabaseCatalog *catalog)
...
+		" ), raw_ordered_fk_constraints as ( "
+		"     select "
+		"            distinct c.confrelid as relid, "
+		"            0 as depth "
+		"       from fk_constraints c "
+		"      where not exists ( "
+		"            select * from "
```
Comment (dimitri) [inline src/bin/pgcopydb/schema.c:434]: Another target list to expand.

Comment (dimitri) [reply src/bin/pgcopydb/schema.c:434]: Also move the FROM clause to its own line?

**Review by dimitri — CHANGES_REQUESTED** [review]
Now I'm nitpicking. Only because it feels like ready to merge, but...

File: src/bin/pgcopydb/schema.c (hunk 4 above)
Comment (JamesGuthrie) [reply src/bin/pgcopydb/schema.c]: done

File: src/bin/pgcopydb/schema.c (hunk 3 above)
Comment (JamesGuthrie) [reply src/bin/pgcopydb/schema.c]: done

**Review by dimitri — APPROVED** [review]

**Comment by dimitri** [comment]
Thanks a lot @JamesGuthrie  for your first contribution to pgcopydb, this is much appreciated!

**Comment by JamesGuthrie** [comment]
Thank you for the swift review!

**Pull request merged**